File: rna_seq_analysis.R
# RNA-seq Data Analysis in R
# This script provides functions for RNA-seq analysis using popular R packages
# including DESeq2, edgeR, and visualization tools

library(DESeq2)
library(edgeR)
library(ggplot2)
library(pheatmap)
library(dplyr)
library(tidyr)

#' Perform DESeq2 Differential Expression Analysis
#'
#' @param count_matrix Count matrix (genes x samples)
#' @param sample_info Data frame with sample information (must include 'condition' column)
#' @param design_formula Design formula (default: ~condition)
#' @param alpha Significance threshold (default: 0.05)
#' @return DESeqDataSet object with results
perform_deseq2_analysis <- function(count_matrix, sample_info, 
                                   design_formula = ~condition, 
                                   alpha = 0.05) {
  
  # Create DESeq2 dataset
  dds <- DESeqDataSetFromMatrix(
    countData = count_matrix,
    colData = sample_info,
    design = design_formula
  )
  
  # Filter low count genes
  keep <- rowSums(counts(dds)) >= 10
  dds <- dds[keep, ]
  
  # Run DESeq2
  dds <- DESeq(dds)
  
  # Get results
  res <- results(dds, alpha = alpha)
  
  # Add gene symbols
  res$gene <- rownames(res)
  
  # Order by p-value
  res <- res[order(res$padj), ]
  
  return(list(dds = dds, results = res))
}


#' Perform edgeR Differential Expression Analysis
#'
#' @param count_matrix Count matrix (genes x samples)
#' @param group_vector Vector of group labels for samples
#' @param fdr FDR threshold (default: 0.05)
#' @return List containing DGEList and results
perform_edger_analysis <- function(count_matrix, group_vector, fdr = 0.05) {
  
  # Create DGEList object
  y <- DGEList(counts = count_matrix, group = group_vector)
  
  # Filter low count genes
  keep <- filterByExpr(y)
  y <- y[keep, , keep.lib.sizes = FALSE]
  
  # Normalize
  y <- calcNormFactors(y)
  
  # Estimate dispersion
  design <- model.matrix(~group_vector)
  y <- estimateDisp(y, design)
  
  # Fit GLM
  fit <- glmQLFit(y, design)
  qlf <- glmQLFTest(fit, coef = 2)
  
  # Get results
  results <- topTags(qlf, n = Inf, adjust.method = "BH")
  
  return(list(dge = y, results = results))
}


#' Normalize counts using TPM
#'
#' @param counts Count matrix
#' @param gene_lengths Named vector of gene lengths
#' @return TPM-normalized matrix
normalize_tpm <- function(counts, gene_lengths) {
  
  # Align gene lengths with count matrix
  gene_lengths <- gene_lengths[rownames(counts)]
  
  # Calculate RPK
  rpk <- counts / (gene_lengths / 1000)
  
  # Calculate scaling factor
  scaling_factor <- colSums(rpk) / 1e6
  
  # Calculate TPM
  tpm <- sweep(rpk, 2, scaling_factor, FUN = "/")
  
  return(tpm)
}


#' Create volcano plot
#'
#' @param results DESeq2 or edgeR results object
#' @param fc_threshold Fold change threshold (default: 2)
#' @param p_threshold P-value threshold (default: 0.05)
#' @param title Plot title
#' @return ggplot object
plot_volcano <- function(results, fc_threshold = 2, p_threshold = 0.05, 
                        title = "Volcano Plot") {
  
  # Convert to data frame if needed
  if (class(results)[1] == "DESeqResults") {
    df <- as.data.frame(results)
    df$log2FoldChange <- df$log2FoldChange
    df$padj <- df$padj
  } else if (class(results)[1] == "TopTags") {
    df <- results$table
    df$log2FoldChange <- df$logFC
    df$padj <- df$FDR
  } else {
    df <- as.data.frame(results)
  }
  
  # Remove NA values
  df <- df[!is.na(df$padj), ]
  
  # Add significance column
  df$significant <- ifelse(
    abs(df$log2FoldChange) >= log2(fc_threshold) & df$padj <= p_threshold,
    "Significant",
    "Not Significant"
  )
  
  # Create plot
  p <- ggplot(df, aes(x = log2FoldChange, y = -log10(padj), color = significant)) +
    geom_point(alpha = 0.5, size = 2) +
    scale_color_manual(values = c("Not Significant" = "gray", "Significant" = "red")) +
    geom_hline(yintercept = -log10(p_threshold), linetype = "dashed", color = "blue") +
    geom_vline(xintercept = c(-log2(fc_threshold), log2(fc_threshold)), 
               linetype = "dashed", color = "green") +
    labs(
      title = title,
      x = "Log2 Fold Change",
      y = "-Log10 Adjusted P-value",
      color = "Status"
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 14, face = "bold"),
      legend.position = "right"
    )
  
  return(p)
}


#' Create MA plot
#'
#' @param results DESeq2 results object
#' @param alpha Significance threshold
#' @return ggplot object
plot_ma <- function(results, alpha = 0.05) {
  
  df <- as.data.frame(results)
  df <- df[!is.na(df$padj), ]
  df$significant <- ifelse(df$padj <= alpha, "Significant", "Not Significant")
  
  p <- ggplot(df, aes(x = baseMean, y = log2FoldChange, color = significant)) +
    geom_point(alpha = 0.5, size = 1) +
    scale_x_log10() +
    scale_color_manual(values = c("Not Significant" = "gray", "Significant" = "red")) +
    geom_hline(yintercept = 0, linetype = "dashed") +
    labs(
      title = "MA Plot",
      x = "Mean Expression (log10)",
      y = "Log2 Fold Change",
      color = "Status"
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 14, face = "bold")
    )
  
  return(p)
}


#' Create expression heatmap
#'
#' @param vsd Variance stabilized or rlog transformed data
#' @param top_genes Number of top genes to include
#' @param annotation_col Data frame with sample annotations
#' @return pheatmap object
plot_expression_heatmap <- function(vsd, top_genes = 50, annotation_col = NULL) {
  
  # Get top variable genes
  rv <- rowVars(assay(vsd))
  select <- order(rv, decreasing = TRUE)[seq_len(min(top_genes, length(rv)))]
  
  # Get data for selected genes
  mat <- assay(vsd)[select, ]
  
  # Z-score normalization
  mat <- t(scale(t(mat)))
  
  # Create heatmap
  pheatmap(
    mat,
    cluster_rows = TRUE,
    cluster_cols = TRUE,
    show_rownames = ifelse(top_genes <= 50, TRUE, FALSE),
    annotation_col = annotation_col,
    color = colorRampPalette(c("blue", "white", "red"))(100),
    breaks = seq(-2, 2, length.out = 101),
    main = paste("Top", top_genes, "Variable Genes"),
    fontsize = 10
  )
}


#' Perform PCA and create plot
#'
#' @param vsd Variance stabilized or rlog transformed data
#' @param intgroup Column name(s) from colData to use for grouping
#' @return ggplot object
plot_pca_analysis <- function(vsd, intgroup = "condition") {
  
  pcaData <- plotPCA(vsd, intgroup = intgroup, returnData = TRUE)
  percentVar <- round(100 * attr(pcaData, "percentVar"))
  
  p <- ggplot(pcaData, aes(x = PC1, y = PC2, color = !!sym(intgroup))) +
    geom_point(size = 4, alpha = 0.8) +
    labs(
      title = "PCA Plot",
      x = paste0("PC1: ", percentVar[1], "% variance"),
      y = paste0("PC2: ", percentVar[2], "% variance")
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 14, face = "bold"),
      legend.position = "right"
    )
  
  return(p)
}


#' Perform Gene Set Enrichment Analysis (GSEA)
#'
#' @param ranked_genes Named vector of genes with ranking metric
#' @param gene_sets List of gene sets
#' @return Enrichment results
perform_gsea <- function(ranked_genes, gene_sets) {
  
  # This is a placeholder - actual GSEA would require fgsea or similar
  # Install fgsea if needed: BiocManager::install("fgsea")
  
  if (!requireNamespace("fgsea", quietly = TRUE)) {
    message("fgsea package not installed. Install with: BiocManager::install('fgsea')")
    return(NULL)
  }
  
  library(fgsea)
  
  # Run GSEA
  fgseaRes <- fgsea(
    pathways = gene_sets,
    stats = ranked_genes,
    minSize = 15,
    maxSize = 500
  )
  
  # Order by p-value
  fgseaRes <- fgseaRes[order(fgseaRes$pval), ]
  
  return(fgseaRes)
}


#' Calculate QC metrics
#'
#' @param count_matrix Count matrix
#' @return Data frame with QC metrics
calculate_qc_metrics <- function(count_matrix) {
  
  qc_df <- data.frame(
    sample = colnames(count_matrix),
    total_reads = colSums(count_matrix),
    genes_detected = colSums(count_matrix > 0),
    median_expression = apply(count_matrix, 2, median),
    mean_expression = colMeans(count_matrix)
  )
  
  # Top 10 genes percentage
  top10_pct <- apply(count_matrix, 2, function(x) {
    sum(sort(x, decreasing = TRUE)[1:10]) / sum(x) * 100
  })
  qc_df$top10_genes_pct <- top10_pct
  
  return(qc_df)
}


#' Extract significant genes
#'
#' @param results DESeq2 results
#' @param padj_threshold Adjusted p-value threshold
#' @param fc_threshold Fold change threshold (linear scale)
#' @return Data frame of significant genes
get_significant_genes <- function(results, padj_threshold = 0.05, fc_threshold = 2) {
  
  sig <- results[
    !is.na(results$padj) & 
    results$padj < padj_threshold & 
    abs(results$log2FoldChange) >= log2(fc_threshold),
  ]
  
  sig <- as.data.frame(sig)
  sig <- sig[order(sig$padj), ]
  
  return(sig)
}


# Print available functions
cat("RNA-seq Analysis Functions Loaded:\n")
cat("- perform_deseq2_analysis()\n")
cat("- perform_edger_analysis()\n")
cat("- normalize_tpm()\n")
cat("- plot_volcano()\n")
cat("- plot_ma()\n")
cat("- plot_expression_heatmap()\n")
cat("- plot_pca_analysis()\n")
cat("- perform_gsea()\n")
cat("- calculate_qc_metrics()\n")
cat("- get_significant_genes()\n")

File: rna_seq_analysis.py
"""
RNA-seq Data Analysis Module

This module provides functions for analyzing RNA sequencing data including:
- Data normalization (TPM, FPKM, CPM)
- Differential expression analysis
- Quality control metrics
- Visualization functions
"""

import pandas as pd
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns


def calculate_tpm(counts, gene_lengths):
    """
    Calculate Transcripts Per Million (TPM) normalization.
    
    Parameters:
    -----------
    counts : pd.DataFrame
        Raw count matrix with genes as rows and samples as columns
    gene_lengths : pd.Series or dict
        Gene lengths in base pairs, indexed by gene names
    
    Returns:
    --------
    pd.DataFrame
        TPM-normalized expression matrix
    """
    # Convert gene_lengths to Series if it's a dict
    if isinstance(gene_lengths, dict):
        gene_lengths = pd.Series(gene_lengths)
    
    # Align gene lengths with counts
    gene_lengths = gene_lengths.reindex(counts.index)
    
    # Calculate RPK (reads per kilobase)
    rpk = counts.div(gene_lengths / 1000, axis=0)
    
    # Calculate scaling factor (per million)
    scaling_factor = rpk.sum(axis=0) / 1e6
    
    # Calculate TPM
    tpm = rpk.div(scaling_factor, axis=1)
    
    return tpm


def calculate_fpkm(counts, gene_lengths, total_reads=None):
    """
    Calculate Fragments Per Kilobase Million (FPKM) normalization.
    
    Parameters:
    -----------
    counts : pd.DataFrame
        Raw count matrix with genes as rows and samples as columns
    gene_lengths : pd.Series or dict
        Gene lengths in base pairs
    total_reads : pd.Series, optional
        Total reads per sample. If None, calculated from counts
    
    Returns:
    --------
    pd.DataFrame
        FPKM-normalized expression matrix
    """
    if isinstance(gene_lengths, dict):
        gene_lengths = pd.Series(gene_lengths)
    
    gene_lengths = gene_lengths.reindex(counts.index)
    
    if total_reads is None:
        total_reads = counts.sum(axis=0)
    
    # FPKM = (counts * 1e9) / (gene_length * total_reads)
    fpkm = counts.mul(1e9).div(gene_lengths, axis=0).div(total_reads, axis=1)
    
    return fpkm


def calculate_cpm(counts):
    """
    Calculate Counts Per Million (CPM) normalization.
    
    Parameters:
    -----------
    counts : pd.DataFrame
        Raw count matrix with genes as rows and samples as columns
    
    Returns:
    --------
    pd.DataFrame
        CPM-normalized expression matrix
    """
    total_counts = counts.sum(axis=0)
    cpm = counts.div(total_counts, axis=1) * 1e6
    return cpm


def filter_low_expression(counts, min_count=10, min_samples=2):
    """
    Filter out genes with low expression across samples.
    
    Parameters:
    -----------
    counts : pd.DataFrame
        Count matrix with genes as rows and samples as columns
    min_count : int
        Minimum count threshold
    min_samples : int
        Minimum number of samples that must meet the count threshold
    
    Returns:
    --------
    pd.DataFrame
        Filtered count matrix
    """
    keep = (counts >= min_count).sum(axis=1) >= min_samples
    return counts[keep]


def calculate_fold_change(treatment, control, pseudocount=1):
    """
    Calculate log2 fold change between treatment and control groups.
    
    Parameters:
    -----------
    treatment : pd.DataFrame or pd.Series
        Expression values for treatment samples
    control : pd.DataFrame or pd.Series
        Expression values for control samples
    pseudocount : float
        Pseudocount to add before log transformation
    
    Returns:
    --------
    pd.Series
        Log2 fold change values
    """
    if isinstance(treatment, pd.DataFrame):
        treatment_mean = treatment.mean(axis=1)
    else:
        treatment_mean = treatment
    
    if isinstance(control, pd.DataFrame):
        control_mean = control.mean(axis=1)
    else:
        control_mean = control
    
    log2fc = np.log2((treatment_mean + pseudocount) / (control_mean + pseudocount))
    
    return log2fc


def perform_ttest(treatment, control):
    """
    Perform t-test for differential expression analysis.
    
    Parameters:
    -----------
    treatment : pd.DataFrame
        Expression values for treatment samples (genes x samples)
    control : pd.DataFrame
        Expression values for control samples (genes x samples)
    
    Returns:
    --------
    pd.DataFrame
        Results containing t-statistic, p-value, and adjusted p-value
    """
    results = []
    
    for gene in treatment.index:
        t_values = treatment.loc[gene].values
        c_values = control.loc[gene].values
        
        t_stat, p_val = stats.ttest_ind(t_values, c_values)
        
        results.append({
            'gene': gene,
            't_statistic': t_stat,
            'p_value': p_val
        })
    
    results_df = pd.DataFrame(results).set_index('gene')
    
    # FDR correction (Benjamini-Hochberg)
    from statsmodels.stats.multitest import multipletests
    _, results_df['p_adjusted'], _, _ = multipletests(
        results_df['p_value'], 
        method='fdr_bh'
    )
    
    return results_df


def plot_volcano(log2fc, pvalues, fc_threshold=1, p_threshold=0.05, 
                 title='Volcano Plot', figsize=(10, 8)):
    """
    Create a volcano plot for differential expression results.
    
    Parameters:
    -----------
    log2fc : pd.Series
        Log2 fold change values
    pvalues : pd.Series
        P-values (will be converted to -log10)
    fc_threshold : float
        Fold change threshold for significance
    p_threshold : float
        P-value threshold for significance
    title : str
        Plot title
    figsize : tuple
        Figure size
    
    Returns:
    --------
    matplotlib.figure.Figure
        Volcano plot figure
    """
    fig, ax = plt.subplots(figsize=figsize)
    
    # Calculate -log10(p-value)
    neg_log10_p = -np.log10(pvalues)
    
    # Determine significance
    significant = (np.abs(log2fc) >= fc_threshold) & (pvalues <= p_threshold)
    
    # Plot
    colors = ['red' if sig else 'gray' for sig in significant]
    ax.scatter(log2fc, neg_log10_p, c=colors, alpha=0.5, s=20)
    
    # Add threshold lines
    ax.axhline(-np.log10(p_threshold), color='blue', linestyle='--', 
               label=f'p = {p_threshold}')
    ax.axvline(fc_threshold, color='green', linestyle='--', 
               label=f'FC = {fc_threshold}')
    ax.axvline(-fc_threshold, color='green', linestyle='--')
    
    ax.set_xlabel('Log2 Fold Change', fontsize=12)
    ax.set_ylabel('-Log10 P-value', fontsize=12)
    ax.set_title(title, fontsize=14)
    ax.legend()
    ax.grid(alpha=0.3)
    
    plt.tight_layout()
    return fig


def plot_heatmap(expression_data, samples_to_plot=None, genes_to_plot=None,
                 title='Expression Heatmap', figsize=(12, 10)):
    """
    Create a heatmap of expression data.
    
    Parameters:
    -----------
    expression_data : pd.DataFrame
        Expression matrix (genes x samples)
    samples_to_plot : list, optional
        Subset of samples to include
    genes_to_plot : list, optional
        Subset of genes to include
    title : str
        Plot title
    figsize : tuple
        Figure size
    
    Returns:
    --------
    matplotlib.figure.Figure
        Heatmap figure
    """
    # Subset data if specified
    data = expression_data.copy()
    if samples_to_plot is not None:
        data = data[samples_to_plot]
    if genes_to_plot is not None:
        data = data.loc[genes_to_plot]
    
    # Z-score normalization for visualization
    data_zscore = (data - data.mean(axis=1).values.reshape(-1, 1)) / data.std(axis=1).values.reshape(-1, 1)
    
    fig, ax = plt.subplots(figsize=figsize)
    sns.heatmap(data_zscore, cmap='RdBu_r', center=0, 
                cbar_kws={'label': 'Z-score'}, ax=ax)
    ax.set_title(title, fontsize=14)
    ax.set_xlabel('Samples', fontsize=12)
    ax.set_ylabel('Genes', fontsize=12)
    
    plt.tight_layout()
    return fig


def calculate_qc_metrics(counts):
    """
    Calculate quality control metrics for RNA-seq data.
    
    Parameters:
    -----------
    counts : pd.DataFrame
        Raw count matrix
    
    Returns:
    --------
    pd.DataFrame
        QC metrics per sample
    """
    qc_metrics = pd.DataFrame()
    
    qc_metrics['total_reads'] = counts.sum(axis=0)
    qc_metrics['genes_detected'] = (counts > 0).sum(axis=0)
    qc_metrics['median_expression'] = counts.median(axis=0)
    qc_metrics['mean_expression'] = counts.mean(axis=0)
    
    # Top 10 genes percentage
    top10_genes = counts.apply(lambda x: x.nlargest(10).sum() / x.sum() * 100, axis=0)
    qc_metrics['top10_genes_pct'] = top10_genes
    
    return qc_metrics


def plot_pca(expression_data, sample_labels=None, title='PCA Plot', figsize=(10, 8)):
    """
    Perform and plot PCA on expression data.
    
    Parameters:
    -----------
    expression_data : pd.DataFrame
        Expression matrix (genes x samples)
    sample_labels : dict or pd.Series, optional
        Labels for samples for coloring
    title : str
        Plot title
    figsize : tuple
        Figure size
    
    Returns:
    --------
    tuple
        (fig, pc_df) - matplotlib figure and PC coordinates DataFrame
    """
    from sklearn.decomposition import PCA
    from sklearn.preprocessing import StandardScaler
    
    # Transpose and standardize
    data_t = expression_data.T
    scaler = StandardScaler()
    data_scaled = scaler.fit_transform(data_t)
    
    # Perform PCA
    n_components = min(10, data_scaled.shape[0], data_scaled.shape[1])
    pca = PCA(n_components=n_components)
    pc_coords = pca.fit_transform(data_scaled)
    
    # Create DataFrame with results
    pc_df = pd.DataFrame(
        pc_coords[:, :2],
        columns=['PC1', 'PC2'],
        index=expression_data.columns
    )
    
    # Plot
    fig, ax = plt.subplots(figsize=figsize)
    
    if sample_labels is not None:
        if isinstance(sample_labels, dict):
            sample_labels = pd.Series(sample_labels)
        pc_df['label'] = sample_labels.reindex(pc_df.index)
        
        for label in pc_df['label'].unique():
            mask = pc_df['label'] == label
            ax.scatter(pc_df.loc[mask, 'PC1'], pc_df.loc[mask, 'PC2'],
                      label=label, alpha=0.7, s=100)
        ax.legend()
    else:
        ax.scatter(pc_df['PC1'], pc_df['PC2'], alpha=0.7, s=100)
    
    variance_exp = pca.explained_variance_ratio_
    ax.set_xlabel(f'PC1 ({variance_exp[0]*100:.1f}% variance)', fontsize=12)
    ax.set_ylabel(f'PC2 ({variance_exp[1]*100:.1f}% variance)', fontsize=12)
    ax.set_title(title, fontsize=14)
    ax.grid(alpha=0.3)
    
    plt.tight_layout()
    return fig, pc_df


if __name__ == "__main__":
    # Example usage
    print("RNA-seq Analysis Module")
    print("Available functions:")
    print("- calculate_tpm()")
    print("- calculate_fpkm()")
    print("- calculate_cpm()")
    print("- filter_low_expression()")
    print("- calculate_fold_change()")
    print("- perform_ttest()")
    print("- plot_volcano()")
    print("- plot_heatmap()")
    print("- calculate_qc_metrics()")
    print("- plot_pca()")
